packages/serverpod_test/lib/src/database_proxy.dart

```diff
@@ -0,0 +1,234 @@
+  }
+
+  @override
+  Future<T> deleteRow<T extends TableRow>(
+    T row, {
+    Transaction? transaction,
+  }) {
+    return _db.deleteRow<T>(row, transaction: transaction);
+  }
+
+  @override
+  Future<List<T>> deleteWhere<T extends TableRow>({
+    required Expression where,
+    Transaction? transaction,
+  }) {
+    return _db.deleteWhere<T>(where: where, transaction: transaction);
+  }
+
+  @override
+  Future<List<T>> find<T extends TableRow>({
+    Expression? where,
+    int? limit,
+    int? offset,
+    Column? orderBy,
+    List<Order>? orderByList,
+    bool orderDescending = false,
+    Transaction? transaction,
+    Include? include,
+  }) {
+    return _db.find<T>(
+      where: where,
+      limit: limit,
+      offset: offset,
+      orderBy: orderBy,
+      orderByList: orderByList,
+      orderDescending: orderDescending,
+      transaction: transaction,
+      include: include,
+    );
+  }
+
+  @override
+  Future<T?> findById<T extends TableRow>(
+    int id, {
+    Transaction? transaction,
+    Include? include,
+  }) {
+    return _db.findById<T>(id, transaction: transaction, include: include);
+  }
+
+  @override
+  Future<T?> findFirstRow<T extends TableRow>({
+    Expression? where,
+    int? offset,
+    Column? orderBy,
+    List<Order>? orderByList,
+    bool orderDescending = false,
+    Transaction? transaction,
+    Include? include,
+  }) {
+    return _db.findFirstRow<T>(
+      where: where,
+      offset: offset,
+      orderBy: orderBy,
+      orderByList: orderByList,
+      orderDescending: orderDescending,
+      transaction: transaction,
+      include: include,
+    );
+  }
+
+  @override
+  Future<List<T>> insert<T extends TableRow>(
+    List<T> rows, {
+    Transaction? transaction,
+  }) {
+    return _db.insert<T>(rows, transaction: transaction);
+  }
+
+  @override
+  Future<T> insertRow<T extends TableRow>(
+    T row, {
+    Transaction? transaction,
+  }) {
+    return _db.insertRow<T>(row, transaction: transaction);
+  }
+
+  @override
+  Future<bool> testConnection() {
+    return _db.testConnection();
+  }
+
+  @override
+  Future<R> transaction<R>(
+    TransactionFunction<R> transactionFunction, {
+    isUserCall = true,
+  }) async {
+    if (!isUserCall || _databaseTestConfig.areRollbacksDisabled) {
+      return _db.transaction(transactionFunction);
+    }
+
+    var localTransaction = _transactionManager.currentTransaction;
+    if (localTransaction == null) {
+      throw StateError('No ongoing transaction.');
+    }
+
+    try {
+      await _transactionManager.addSavePoint(lock: true);
+    } on ConcurrentTransactionsException {
+      throw InvalidConfigurationException(
+        'Several calls to `transaction` was made concurrently which is not supported with the current database test configuration. '
+        'Disable rolling back the database by setting `databaseTestConfig` to `DatabaseTestConfig.rollbacksDisabled()`.',
```

Comment:
Suggestion: Rephrase to: "Several concurrent calls to `transaction` are not supported when database rollbacks are enabled. Disable rolling back the database by setting... to continue. This makes you responsible for restoring the database after the test as complete.".

I would like to be more distinct on what exactly is causing the exception, what action needs to be taken and possibly the consequence of this action.
Maybe the consequence could be left out of the message since it should be clear when configuring the flag?

Reply:
Good input, that is better 👍